src/scores/processing/block_bootstrap.py

```diff
@@ -0,0 +1,410 @@
+        length
+        """
+        if block == length:
+            return list(range(length))
+        repeats = math.ceil(length / block)
+        if circular:
+            indices = list(
+                chain.from_iterable(
+                    islice(cycle(range(length)), s, s + block) for s in np.random.randint(0, length, repeats)
+                )
+            )
+        else:
+            indices = list(
+                chain.from_iterable(
+                    islice(range(length), s, s + block) for s in np.random.randint(0, length - block + 1, repeats)
+                )
+            )
+        return indices[:length]
+
+    # Don't randomise within an outer block
+    if len(prev_block_sizes) > 0:
+        orig_shape = shape.copy()
+        for i, b in enumerate(prev_block_sizes[::-1]):
+            prev_ax = block_axis - (i + 1)
+            shape[prev_ax] = math.ceil(shape[prev_ax] / b)
+
+    if block_size == 1:
+        indices = np.random.randint(
+            0,
+            shape[block_axis],
+            shape,
+        )
+    else:
+        non_block_shapes = [s for i, s in enumerate(shape) if i != block_axis]
+        indices = np.moveaxis(
+            np.stack(
+                [_random_blocks(shape[block_axis], block_size, circular) for _ in range(np.prod(non_block_shapes))],
+                axis=-1,
+            ).reshape([shape[block_axis]] + non_block_shapes),
+            0,
+            block_axis,
+        )
+
+    if len(prev_block_sizes) > 0:
+        for i, b in enumerate(prev_block_sizes[::-1]):
+            prev_ax = block_axis - (i + 1)
+            indices = np.repeat(indices, b, axis=prev_ax).take(range(orig_shape[prev_ax]), axis=prev_ax)
+        return indices
+    return indices
+
+
+def _n_nested_blocked_random_indices(
+    sizes: OrderedDict[str, Tuple[int, int]], n_iteration: int, circular: bool = True
+) -> OrderedDict[str, np.ndarray]:
+    """
+    Returns indices to randomly resample blocks of an array (with replacement)
+    in a nested manner many times. Here, "nested" resampling means to randomly
+    resample the first dimension, then for each randomly sampled element along
+    that dimension, randomly resample the second dimension, then for each
+    randomly sampled element along that dimension, randomly resample the third
+    dimension etc.
+
+    Args:
+    sizes: Dictionary with {names: (sizes, blocks)} of the dimensions to resample
+    n_iteration: The number of times to repeat the random resampling
+    circular: Whether or not to do circular resampling.
+
+    Returns:
+        A dictionary of arrays containing indices for nested block resampling.
+
+    """
+
+    shape = [s[0] for s in sizes.values()]
+    indices = OrderedDict()
+    prev_blocks: List[int] = []
+    for ax, (key, (_, block)) in enumerate(sizes.items()):
+        indices[key] = _get_blocked_random_indices(shape[: ax + 1] + [n_iteration], ax, block, prev_blocks, circular)
```

Comment:
Ok so, a combination of indexing `shape[: ax + 1]` and maybe `+ [n_iteration]` seems to create a copy of `shape` (because it's a list). Therefore, even if it is manipulated in `_get_blocked_random_indices` (see previous comment), the slicing creates unique obj each time, and hence the function doesn't mutate the original `shape`.

However, this wasn't obvious to me initially. Again it may be good to make this behaviour explicit by passing in a "frozen" list or similar, and/or instead of creating a `.copy()` in `_get_blocked_random_indices` and manipulating the original, manipulate the copy instead.